File: src/main/kotlin/ai/astha/scantheplanet/idea/scanner/PromptTokenEstimator.kt
package ai.astha.scantheplanet.idea.scanner

import ai.astha.scantheplanet.idea.scanner.prompt.CodeChunkPrompt
import ai.astha.scantheplanet.idea.scanner.prompt.PromptBuilder
import ai.astha.scantheplanet.idea.scanner.prompt.PromptPayload
import ai.astha.scantheplanet.idea.scanner.prompt.MitigationPrompt
import com.knuddels.jtokkit.Encodings
import com.knuddels.jtokkit.api.Encoding
import com.knuddels.jtokkit.api.EncodingRegistry
import com.knuddels.jtokkit.api.EncodingType

class PromptTokenEstimator(providerName: String, modelName: String?) {
    private val provider = providerName.lowercase()
    private val encoding = encodingForModel(modelName)

    fun baseTokens(
        technique: TechniqueSpec,
        readmeExcerpt: String?,
        mitigations: List<MitigationPrompt>,
        filePath: String
    ): Int {
        val payload = PromptPayload(
            techniqueId = technique.id,
            techniqueName = technique.name,
            severity = technique.severity,
            summary = technique.summary,
            description = technique.description,
            mitigations = mitigations,
            codeChunk = CodeChunkPrompt(
                id = "base",
                file = filePath,
                startLine = 0,
                endLine = 0,
                code = "",
                ruleHints = emptyList()
            ),
            readmeExcerpt = readmeExcerpt
        )
        val userPrompt = PromptBuilder.buildUserPrompt(payload)
        var tokens = countTokens(userPrompt)
        tokens += countTokens(PromptBuilder.SYSTEM_PROMPT)
        if (provider == "anthropic") {
            tokens += countTokens(PromptBuilder.JSON_PREFILL)
        }
        tokens += messageOverheadTokens(provider)
        return tokens
    }

    fun countTokens(text: String): Int {
        return encoding.countTokensOrdinary(text)
    }

    private fun messageOverheadTokens(provider: String): Int {
        return when (provider) {
            "openai" -> 12
            "anthropic" -> 12
            "gemini" -> 8
            else -> 10
        }
    }

    companion object {
        private val registry: EncodingRegistry = Encodings.newLazyEncodingRegistry()

        private fun encodingForModel(modelName: String?): Encoding {
            val resolved = modelName?.let { registry.getEncodingForModel(it) }
            if (resolved != null && resolved.isPresent) {
                return resolved.get()
            }
            val fallback = when {
                modelName != null && modelName.lowercase().contains("gpt-4o") -> EncodingType.O200K_BASE
                modelName != null && modelName.lowercase().startsWith("gpt-5") -> EncodingType.O200K_BASE
                else -> EncodingType.CL100K_BASE
            }
            return registry.getEncoding(fallback)
        }
    }
}

File: packages/core/src/scanner/PromptTokenEstimator.ts
import type { TechniqueSpec } from '@scantheplanet/safe-mcp';
import type { MitigationPrompt } from '../prompt/PromptModels.js';
import { buildUserPrompt, SYSTEM_PROMPT, JSON_PREFILL } from '../prompt/PromptBuilder.js';
import type { CodeChunkPrompt, PromptPayload } from '../prompt/PromptModels.js';

export class PromptTokenEstimator {
  constructor(private readonly providerName: string, private readonly modelName?: string) {}

  baseTokens(
    technique: TechniqueSpec,
    readmeExcerpt: string | null,
    mitigations: MitigationPrompt[],
    filePath: string
  ): number {
    const payload: PromptPayload = {
      techniqueId: technique.id,
      techniqueName: technique.name,
      severity: technique.severity,
      summary: technique.summary,
      description: technique.description,
      mitigations,
      codeChunk: {
        id: 'base',
        file: filePath,
        startLine: 0,
        endLine: 0,
        code: '',
        ruleHints: [],
      } as CodeChunkPrompt,
      readmeExcerpt: readmeExcerpt ?? undefined,
    };
    const userPrompt = buildUserPrompt(payload);
    let tokens = this.countTokens(userPrompt);
    tokens += this.countTokens(SYSTEM_PROMPT);
    if (this.providerName.toLowerCase() === 'anthropic') {
      tokens += this.countTokens(JSON_PREFILL);
    }
    tokens += messageOverheadTokens(this.providerName);
    return tokens;
  }

  countTokens(text: string): number {
    // Lightweight approximation; avoids heavy tokenizer dependency.
    const normalized = text.replace(/\s+/g, ' ').trim();
    if (!normalized) return 0;
    return Math.max(1, Math.ceil(normalized.length / 4));
  }
}

function messageOverheadTokens(provider: string): number {
  const lower = provider.toLowerCase();
  switch (lower) {
    case 'openai':
      return 12;
    case 'anthropic':
      return 12;
    case 'gemini':
      return 8;
    default:
      return 10;
  }
}
